File: src/main/java/META-INF/resources/dataforms/app/user/page/PasswordRecenryptForm.js
/**
 * @fileOverview {@link PasswordRecenryptForm}クラスを記述したファイルです。
 */

'use strict';

import { JsonResponse } from "../../../response/JsonResponse.js";
import { MessagesUtil } from "../../../util/MessagesUtil.js";
import { ValidationError } from '../../../validator/ValidationError.js';

/**
 * @class PasswordRecenryptForm
 *
 * @extends Form
 */
export class PasswordRecenryptForm extends Form {
	/**
	 * コンストラクタ。
	 */
	constructor() {
		super();
	}

	/**
	 * HTMLエレメントとの対応付けを行います。
	 */
	attach() {
		super.attach();
		this.get("reencryptButton").click(() => {
			this.reencrypt();
		});
	}

	/**
	 * フォームのバリデーション。
	 */
	validateForm() {
		let list = super.validateForm();
		if (list.length == 0) {
			let p = this.get("cryptPassword").val();
			let pc = this.get("cryptPasswordCheck").val();
			if (p != pc) {
				list.push(new ValidationError("cryptPasswordCheck", MessagesUtil.getMessage("error.passwordnotmatch")));
			}
		}
		return list;
	}

	/**
	 * 再暗号化処理。
	 */
	async reencrypt() {
		try {
			if (this.validate()) {
				this.parent.resetErrorStatus();
				if (await currentPage.confirm(null, MessagesUtil.getMessage("message.confirmreencrypted"))) {
					let r = await this.submit("reencrypt");
					this.parent.resetErrorStatus();
					if (r.status == JsonResponse.INVALID) {
						currentPage.setErrorInfo(this.getValidationResult(r), this);
					} else {
						currentPage.alert(null, r.result);
					}
				}
			}
		} catch (e) {
			currentPage.reportError(e);
		}
	}
}

File: src/main/java/jp/dataforms/fw/app/user/page/PasswordRecenryptForm.java
package jp.dataforms.fw.app.user.page;

import java.util.List;
import java.util.Map;

import jp.dataforms.fw.annotation.WebMethod;
import jp.dataforms.fw.app.user.dao.UserDao;
import jp.dataforms.fw.app.user.field.CryptAlgolithmField;
import jp.dataforms.fw.app.user.field.PasswordField;
import jp.dataforms.fw.controller.Form;
import jp.dataforms.fw.response.JsonResponse;
import jp.dataforms.fw.response.Response;
import jp.dataforms.fw.util.MessagesUtil;
import jp.dataforms.fw.validator.RequiredValidator;
import jp.dataforms.fw.validator.ValidationError;

/**
 * 問い合わせフォームクラス。
 */
public class PasswordRecenryptForm extends Form {
	/**
	 * 暗号化パスワードフィールドID。
	 */
	private static final String ID_CRYPT_PASSWORD = "cryptPassword";
	/**
	 * 暗号化パスワード(確認)フィールドID。
	 */
	private static final String ID_CRYPT_PASSWORD_CHECK = "cryptPasswordCheck";

	/**
	 * コンストラクタ。
	 */
	public PasswordRecenryptForm() {
		super(null);
		this.addField(new CryptAlgolithmField()).setComment("暗号化アルゴリズム").addValidator(new RequiredValidator());
		this.addField(new PasswordField(ID_CRYPT_PASSWORD)).setComment("パスワード/キー").addValidator(new RequiredValidator());
		this.addField(new PasswordField(ID_CRYPT_PASSWORD_CHECK)).setComment("パスワード/キー(確認)").addValidator(new RequiredValidator());
	}

	@Override
	protected List<ValidationError> validateForm(Map<String, Object> data) throws Exception {
		List<ValidationError> list = super.validateForm(data);
		if (list.size() == 0) {
			String password = (String) data.get(ID_CRYPT_PASSWORD);
			String passwordCheck = (String) data.get(ID_CRYPT_PASSWORD_CHECK);
			if (!password.equals(passwordCheck)) {
				String msg = MessagesUtil.getMessage(this.getPage(), "error.passwordnotmatch");
				list.add(new ValidationError(ID_CRYPT_PASSWORD_CHECK, msg));
			}
		}
		return list;
	}

	/**
	 * パスワードを更新します。
	 * @param data データ。
	 * @throws Exception 例外。
	 */
	private void updatePassword(final Map<String, Object> data) throws Exception {
		UserDao dao = new UserDao(this);
		String algolithm = (String) data.get("cryptAlgolithm");
		String password = (String) data.get(ID_CRYPT_PASSWORD);
		dao.reencryptPassword(algolithm, password);
	}

	/**
	 * 再暗号化を行います。
	 * @param p パラメータ。
	 * @return 処理結果。
	 * @throws Exception 例外。
	 */
	@WebMethod
	public Response reencrypt(final Map<String, Object> p) throws Exception {
		Response resp = null;
		List<ValidationError> list = this.validate(p);
		if (list.size() == 0) {
			Map<String, Object> data = this.convertToServerData(p);
			this.updatePassword(data);
			resp = new JsonResponse(JsonResponse.SUCCESS, MessagesUtil.getMessage(this.getPage(), "message.passwordreencrypted"));
		} else {
			resp = new JsonResponse(JsonResponse.INVALID, list);
		}
		return resp;
	}
}
